worker.py

```diff
@@ -56,152 +50,64 @@
-                                      'position': rnd.uniform(*spawn_region),
-                                      'linear_velocity': rnd.uniform(*velocity_range)}))
-
-# --- load models & place them in the simulator
-simulator = Simulator(frame_rate=FLAGS.frame_rate, step_rate=FLAGS.step_rate)
-simulator.add(floor)
+objects = [asset_source.get_random_object(rnd) for _ in range(nr_objects)]
+
 for obj in objects:
   collision = simulator.add(obj)
-  # TODO do something about detected collisions
+  trial = 0
+  while collision and trial < 100:
+    obj.position, obj.rotation = asset_source.get_random_object_pose(obj.name, rnd)
+    collision = simulator.check_overlap(obj)
+    trial += 1
+  if collision:
+    raise RuntimeError('Failed to place', obj)
 
 # --- run the physics simulation
-animation = simulator.run(FLAGS.frame_end)
+animation = simulator.run()
 
 # ------------------------------------------------------------------------------
 # ------------------------------------------------------------------------------
 # ------------------------------------------------------------------------------
-
-# --- set up the rendering
-renderer = THREE.Renderer()
-renderer.set_size(FLAGS.resolution, FLAGS.resolution)
-scene = THREE.Scene()
-scene.frame_start = FLAGS.frame_start
-scene.frame_end = FLAGS.frame_end
-renderer.set_up_background(bg_color=(0., 0., 0., 0.))
-renderer.set_up_exr_output(path=FLAGS.outpath)
-
-# --- Environment settings for CLEVR
-
-# --- Camera settings from CLEVR
-camera = THREE.PerspectiveCamera(focal_length=35.)
-camera.position = (7.48113, -6.50764, 5.34367)
-camera.quaternion = (0.7816, 0.481707, 0.212922, 0.334251)
-camera.camera.sensor_width = 32
-
-# --- Light settings from CLEVR
-sun = THREE.DirectionalLight(color=0xffffff, intensity=0.45, shadow_softness=0.2)
-sun.position = (11.6608, -6.62799, 25.8232)
-sun.quaternion = (0.971588, 0.105085, 0.210842, 0.022804)
-scene.add(sun)
-lamp_back = THREE.RectAreaLight(color=0xffffff, intensity=50., width=1, height=1)
-lamp_back.position = (-1.1685, 2.64602, 5.81574)
-lamp_back.look_at(0, 0, 0)
-scene.add(lamp_back)
-lamp_key = THREE.RectAreaLight(color=0xffedd0, intensity=100, width=0.5, height=0.5)
-lamp_key.position = (6.44671, -2.90517, 4.2584)
-lamp_key.look_at(0, 0, 0)
-scene.add(lamp_key)
-lamp_fill = THREE.RectAreaLight(color=0xc2d0ff, intensity=30, width=0.5, height=0.5)
-lamp_fill.position = (-4.67112, -4.0136, 3.01122)
-lamp_fill.look_at(0, 0, 0)
-scene.add(lamp_fill)
-
-
-# TODO: this is a hack; conversion should be done automatically.
-def translate_quat(pb_quat):
-  """ Convert pyBullet XYZW quaternions into Blender WXYZ quaternions."""
-  x, y, z, w = pb_quat
-  return [w, x, y, z]
-
-
-# --- Dump the simulation data in the renderer
-floor_mesh = THREE.Mesh.from_file(str(floor.vis_filename))
-floor_mesh.position = floor.position
-floor_mesh.quaternion = translate_quat(floor.rotation)
-scene.add(floor_mesh)
+renderer = Blender(scene)
```

Comment:
Haha, c'mon! I moved this line down, **because you asked me to**. 😆 
(You wanted to group all the renderer calls together at the bottom instead of scattering them across the file.)

Yes you can call this as soon as the scene is created if you like, and that can be used for real-time prototyping. Since (in Kubric) you need the `Blender` renderer to create a `.blend` file, there will obviously be none until after you do this. But even after this line it will be empty until you've added the objects too.

Reply:
Ahah, we discussed about this in person, it was not about the sorting in `worker.py` :)
(2nd part of your answer addressed this)